core/order.go

```diff
@@ -380,7 +390,15 @@ func (n *OpenBazaarNode) Purchase(data *PurchaseData) (orderId string, paymentAd
 			return orderId, contract.BuyerOrder.Payment.Address, contract.BuyerOrder.Payment.Amount, false, err
 		} else { // Vendor responded
 			if resp.MessageType == pb.Message_ERROR {
-				return "", "", 0, false, fmt.Errorf("Vendor rejected order, reason: %s", string(resp.Payload.Value))
+				var errStr string
+				errMsg := new(pb.Error)
+				err := ptypes.UnmarshalAny(resp.Payload, errMsg)
+				if err == nil {
+					errStr = errMsg.ErrorMessage
+				} else { // For backwards compatibility check for a string payload
+					errStr = string(resp.Payload.Value)
+				}
+				return "", "", 0, false, fmt.Errorf("Vendor rejected order, reason: %s", errStr)
```

Comment:
The second time you write the same exact code should be an indication that it should be extracted into its own function.

`str, err := extractErrorString(resp)`